tests/string_test.c

```diff
@@ -0,0 +1,310 @@
+    expected = "";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr, value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_ends_with_token, test_char_split_ends_with_token_fn)
+
+static int test_char_split_begins_with_token_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = ";testa;testb;testc";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&output, allocator, 4, sizeof(struct aws_byte_cursor)));
+    ASSERT_SUCCESS(aws_string_split_on_char(&to_split, ';', &output));
+    ASSERT_INT_EQUALS(4, aws_array_list_length(&output));
+
+    struct aws_byte_cursor value = {0};
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 0));
+    char *expected = "";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 1));
+    expected = "testa";
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 2));
+    expected = "testb";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 3));
+    expected = "testc";
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_begins_with_token, test_char_split_begins_with_token_fn)
+
+static int test_char_split_token_not_present_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = "testa";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&output, allocator, 4, sizeof(struct aws_byte_cursor)));
+    ASSERT_SUCCESS(aws_string_split_on_char(&to_split, ';', &output));
+    ASSERT_INT_EQUALS(1, aws_array_list_length(&output));
+
+    struct aws_byte_cursor value = {0};
+    ASSERT_SUCCESS(aws_array_list_get_at(&output, &value, 0));
+
+    char *expected = "testa";
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected, strlen(expected), value.ptr,  value.len);
+
+    aws_array_list_clean_up(&output);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_char_split_token_not_present, test_char_split_token_not_present_fn)
+
+static int test_char_split_empty_fn(struct aws_allocator *allocator, void *ctx) {
+    const char str_to_split[] = "";
+
+    struct aws_byte_buf to_split = aws_byte_buf_from_literal(str_to_split);
+
+    struct aws_array_list output;
+    ASSERT_SUCCESS(aws_array_list_init_dynamic(&output, allocator, 4, sizeof(struct aws_byte_cursor)));
+    ASSERT_SUCCESS(aws_string_split_on_char(&to_split, ';', &output));
+    ASSERT_INT_EQUALS(0, aws_array_list_length(&output));
```

Comment:
This behavior seems a bit inconsistent. I would argue that the number of elements in the output should be equal to the number of splits plus one. Having a weird edge case where it's zero seems like something that could trip someone up and lead to NPEs and similar.

Reply:
I like this, this also removes a branch. Will do